### PR TITLE
Add dl_iterate_phdr to some BSD OSs

### PR DESCRIPTION
**Pull request by lzutao**
Add dl_iterate_phdr for freebsd, openbsd and netbsd: #1066
cc https://github.com/rust-lang/backtrace-rs/issues/325

### CODE REVIEW

**Comment by lzutao** [comment]
There is overlapping C macro expansion: https://c.godbolt.org/z/zT9PE1

Waiting for helping ! :)

**Comment by lzutao** [comment]
Done. CI should be green!

**Comment by semarie** [comment]
I just tested it on OpenBSD, and it failed due to `p_type` error. Adding a `cfg.skip_field` section as for freebsd makes the tests to pass on OpenBSD.

Could you add it inside test_openbsd() ? Else I will correct it after your PR to be commited.

Thanks to have provided the types and function on OpenBSD !

**Comment by semarie** [comment]
tests are fine on OpenBSD. thanks !

**Comment by lzutao** [comment]
Looks like it works: <https://github.com/rust-lang/rust/issues/78184#issuecomment-716859849>.
Ping @JohnTitor @joshtriplett

File: src/unix/bsd/netbsdlike/netbsd/mod.rs [hunk 1]
```diff
@@ -11,6 +11,36 @@ pub type vm_size_t = ::uintptr_t;
...
+
+cfg_if! {
+    if #[cfg(target_pointer_width = "64")] {
+        type Elf_Addr = Elf64_Addr;
+        type Elf_Half = Elf64_Half;
+        type Elf_Phdr = Elf64_Phdr;
+    } else if #[cfg(target_pointer_width = "32")] {
+        type Elf_Addr = Elf32_Addr;
+        type Elf_Half = Elf32_Half;
+        type Elf_Phdr = Elf32_Phdr;
```
Comment (JohnTitor) [inline src/unix/bsd/netbsdlike/netbsd/mod.rs:40]:
```suggestion
        pub type Elf_Addr = Elf64_Addr;
        pub type Elf_Half = Elf64_Half;
        pub type Elf_Phdr = Elf64_Phdr;
    } else if #[cfg(target_pointer_width = "32")] {
        pub type Elf_Addr = Elf32_Addr;
        pub type Elf_Half = Elf32_Half;
        pub type Elf_Phdr = Elf32_Phdr;
```

Shouldn't be?

Comment (lzutao) [reply src/unix/bsd/netbsdlike/netbsd/mod.rs:40]: This type alias is implementation details. In manpages, it uses C macros instead.

**Comment by JohnTitor** [comment]
Looks good to me, thanks!
@bors r+

**Comment by JohnTitor** [comment]
The failure is:
```
error: type alias is never used: `Elf_Addr`
  --> src/unix/bsd/netbsdlike/netbsd/mod.rs:34:9
   |
34 |         type Elf_Addr = Elf64_Addr;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/lib.rs:18:38
   |
18 | #![cfg_attr(libc_deny_warnings, deny(warnings))]
   |                                      ^^^^^^^^
   = note: `#[deny(dead_code)]` implied by `#[deny(warnings)]`

error: type alias is never used: `Elf_Half`
  --> src/unix/bsd/netbsdlike/netbsd/mod.rs:35:9
   |
35 |         type Elf_Half = Elf64_Half;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: type alias is never used: `Elf_Phdr`
  --> src/unix/bsd/netbsdlike/netbsd/mod.rs:36:9
   |
36 |         type Elf_Phdr = Elf64_Phdr;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors
```

Ah, anyway we should "use" them here.

**Comment by lzutao** [comment]
I'm sorry. I made a mistake in netbsd code. The diff of change is 5c23d7574a40fe6243a2207f7938addec91d8693..be37f011e698457058a882d29c34640bbc73653d .

**Comment by lzutao** [comment]
What is this error? <https://github.com/rust-lang/libc/runs/1314551399#step:5:1496>
```
2020-10-27T12:32:07.6255772Z LLVM ERROR: unable to allocate function return #6
2020-10-27T12:32:07.6305138Z error: could not compile `compiler_builtins`
```

**Comment by JohnTitor** [comment]
The second error is unrelated, `sparc-unknown-linux-gnu` is no core target and we use the `build-std` feature in that case.
I think the first error is the cause, I recently updated `compiler_builtins` to 0.1.36 (https://github.com/rust-lang/rust/pull/78209) and it seems that the bump causes the failure.

**Comment by JohnTitor** [comment]
@bors retry
Encountered the spurious failure on rollup :(

**Comment by lzutao** [comment]
Can we create a new patch release for this PR ?

**Pull request merged**